File: downloads/code/LeetCode/Java/House-Robber3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int rob(TreeNode root) {
        if(root==null) return 0;
        int[] ret = dfs(root);    
        return Math.max(ret[0],ret[1]);
    }

    public int[] dfs(TreeNode node){
        // ret[0]: rob, ret[1]: non-rob
        int[] ret = new int[2];
        
        if(node==null){
            ret[0] = 0;
            ret[1] = 0;
            return ret;
        }

        int[] left = dfs(node.left);
        int[] right = dfs(node.right);

        ret[0] = left[1]+right[1]+node.val;
        ret[1] = Math.max(left[0],left[1])+Math.max(right[1],right[0]);

        return ret;
    }

}

File: downloads/code/LeetCode/Python/House-Robber3.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):

    # DFS
    # return: 抢劫node的收益， 不抢劫node的受益
    def dfs(self,node):
        if node == None:
            return 0,0
        rob_L, no_rob_L = self.dfs(node.left)
        rob_R, no_rob_R = self.dfs(node.right)
        return no_rob_L+no_rob_R+node.val, max(no_rob_R,rob_R) + max(no_rob_L, rob_L)
    

    def rob(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """

        ret = self.dfs(root)
        return max(ret[0],ret[1])
